in_toto/verifylib.py

```diff
@@ -889,16 +879,13 @@ def verify_delete_rule(rule, source_materials_queue, source_products_queue):
   """
   rule_data = in_toto.rulelib.unpack_rule(rule)
 
-  matched_materials = fnmatch.filter(
+  filtered_materials = fnmatch.filter(
       source_materials_queue, rule_data["pattern"])
 
-  for matched_material in matched_materials:
-    if matched_material in source_products_queue:
-      raise RuleVerificationError("Rule '{0}' failed, material '{1}' was found"
-          " in products but should have been deleted."
-              .format(" ".join(rule), matched_material))
+  filtered_products = fnmatch.filter(
+      source_products_queue, rule_data["pattern"])
 
-  return list(set(source_materials_queue) - set(matched_materials))
+  return set(filtered_materials) - set(filtered_products)
```

Comment:
I'm still surprised everywhere else this is treated like a set and nothing happened...

Reply:
Not sure I understand. Before, all `verify_*_rule` functions returned the updated queue(s) as `list` (except for `disallow`). Then I discovered that `match` also modified the queue at the passed reference, which is a bug that I fixed in 4b3028aab6b83823b1c8c78b7a8e4d9244b5e554.

And then I realized that there are other things in the way the queues are updated that need fixing, which I've started in bf0709bb96f3f5f2eeac8723c1146b627bdc84c5, in the course of which I'm trying to simplify the code and align it more with the pseudo code of the spec, e.g. having `verify_*_rule` return the consumed artifacts (as `set`) and update the queue(s) on the caller side.

The reason why we haven't noticed is probably that it requires very specific combinations and sequences of rules and artifact queues to become an issue (and probably because the code is confusing).

I'll follow up with a more precise description of the issue and thorough tests.